NeutronStandard/Sniffs/Functions/TypeHintSniff.php

```diff
@@ -30,53 +30,61 @@ private function checkForMissingArgumentHints(File $phpcsFile, $stackPtr) {
 
 		for ($ptr = ($openParenPtr + 1); $ptr < $closeParenPtr; $ptr++) {
 			if ($tokens[$ptr]['code'] === T_VARIABLE) {
-				$tokenBeforePtr = $this->helper->getArgumentTypePtr($phpcsFile, $ptr);
+				$tokenBeforePtr = $helper->getArgumentTypePtr($phpcsFile, $ptr);
 				$tokenBefore = $tokens[$tokenBeforePtr];
-				if (! $tokenBeforePtr || ! in_array($tokenBefore['code'], $hintTypes, true)) {
+				if (!$tokenBeforePtr || !in_array($tokenBefore['code'], $hintTypes, true)) {
 					$error = 'Argument type is missing';
 					$phpcsFile->addWarning($error, $stackPtr, 'NoArgumentType');
 				}
 			}
 		}
 	}
 
-	private function checkForMissingReturnHints(File $phpcsFile, $stackPtr) {
+	private function checkForMissingReturnHints(File $phpcsFile, $stackPtr, SniffHelpers $helper) {
 		$tokens = $phpcsFile->getTokens();
-		if ($this->helper->isFunctionJustSignature($phpcsFile, $stackPtr)) {
+		if ($helper->isFunctionJustSignature($phpcsFile, $stackPtr)) {
 			return;
 		}
-		$endOfFunctionPtr = $this->helper->getEndOfFunctionPtr($phpcsFile, $stackPtr);
-		$startOfFunctionPtr = $this->helper->getStartOfFunctionPtr($phpcsFile, $stackPtr);
-		$returnTypePtr = $this->helper->getNextReturnTypePtr($phpcsFile, $stackPtr);
+		$endOfFunctionPtr = $helper->getEndOfFunctionPtr($phpcsFile, $stackPtr);
+		$startOfFunctionPtr = $helper->getStartOfFunctionPtr($phpcsFile, $stackPtr);
+		$returnTypePtr = $helper->getNextReturnTypePtr($phpcsFile, $stackPtr);
 		$returnType = $tokens[$returnTypePtr];
 
-		$foundReturn = false;
+		$nonVoidReturn = 0;
+		$voidReturn = 0;
 		$scopeClosers = [];
 		for ($ptr = $startOfFunctionPtr; $ptr < $endOfFunctionPtr; $ptr++) {
 			$token = $tokens[$ptr];
-			if (! empty($scopeClosers) && $ptr === $scopeClosers[0]) {
+			if (!empty($scopeClosers) && $ptr === $scopeClosers[0]) {
 				array_shift($scopeClosers);
 			}
 			if ($token['code'] === T_CLOSURE) {
 				array_unshift($scopeClosers, $token['scope_closer']);
 			}
-			if (empty($scopeClosers) && $token['code'] === T_RETURN
-				&& ! $this->helper->isReturnValueVoid($phpcsFile, $ptr)) {
-				$foundReturn = true;
+			if (empty($scopeClosers) && $token['code'] === T_RETURN) {
+				$helper->isReturnValueVoid($phpcsFile, $ptr) ? $voidReturn++ : $nonVoidReturn++;
 			}
 		}
 
-		if (! $foundReturn && $returnTypePtr && $returnType['content'] !== 'void') {
-			$error = 'Return type with no return';
-			$phpcsFile->addWarning($error, $stackPtr, 'UnusedReturnType');
+		if ($returnTypePtr && $returnType['content'] !== 'void') {
+			if (!$nonVoidReturn) {
+				$error = 'Return type with no return';
+				$phpcsFile->addError($error, $stackPtr, 'UnusedReturnType');
+
+				return;
+			}
+			if ($voidReturn) {
```

Comment:
Since these are integers, could we explicitly make the comparisons integer comparisons rather than a truthy check? So this would be `if ($voidReturn > 0)` and the ones below would be `if ($nonVoidReturn > 0 && ...`. I know that a truthy check is the same result here, but I worry that in the future it might trick me into accidentally assuming one of these is a boolean, causing a similar problem to the one this PR is designed to solve. 😁

Reply:
If they are called `$voidReturnCount` as you suggested should not be wrongly asumed as boolean, but if you want to have that, it's easy and does not hurt.